packages/remix-react/links.ts

```diff
@@ -282,60 +285,31 @@ export async function getStylesheetPrefetchLinks(
 // This is ridiculously identical to transition.ts `filterMatchesToLoad`
 export function getNewMatchesForLinks(
   page: string,
-  nextMatches: RouteMatch<ClientRoute>[],
-  currentMatches: RouteMatch<ClientRoute>[],
+  nextMatches: ClientMatch[],
+  currentMatches: ClientMatch[],
   location: Location,
   mode: "data" | "assets"
-): RouteMatch<ClientRoute>[] {
+): ClientMatch[] {
   let path = parsePathPatch(page);
 
-  let isNew = (match: RouteMatch<ClientRoute>, index: number) => {
-    if (!currentMatches[index]) return true;
-    return match.route.id !== currentMatches[index].route.id;
-  };
-
-  let matchPathChanged = (match: RouteMatch<ClientRoute>, index: number) => {
-    return (
-      // param change, /users/123 -> /users/456
-      currentMatches[index].pathname !== match.pathname ||
-      // splat param changed, which is not present in match.path
-      // e.g. /files/images/avatar.jpg -> files/finances.xls
-      (currentMatches[index].route.path?.endsWith("*") &&
-        currentMatches[index].params["*"] !== match.params["*"])
-    );
-  };
+  let prevUrl = new URL(
+    location.pathname + location.search + location.hash,
+    window.origin
+  );
+  let url = new URL(page, window.origin);
+  let filterByRouteProps = filterByRoutePropsFactory(
+    currentMatches,
+    prevUrl,
+    url
+  );
 
-  // NOTE: keep this mostly up-to-date w/ the transition data diff, but this
-  // version doesn't care about submissions
   let newMatches =
     mode === "data" && location.search !== path.search
-      ? // this is really similar to stuff in transition.ts, maybe somebody smarter
-        // than me (or in less of a hurry) can share some of it. You're the best.
-        nextMatches.filter((match, index) => {
-          if (!match.route.hasLoader) {
-            return false;
-          }
-
-          if (isNew(match, index) || matchPathChanged(match, index)) {
-            return true;
-          }
-
-          if (match.route.shouldReload) {
-            return match.route.shouldReload({
-              params: match.params,
-              prevUrl: new URL(
-                location.pathname + location.search + location.hash,
-                window.origin
-              ),
-              url: new URL(page, window.origin),
-            });
-          }
-          return true;
-        })
```

Comment:
The filter function here is identical between both usages, requiring `currentMatches`, `prevUrl`, `url`, and optionally `submission`.  Went with a factory to put them in a closure for now, but this could also be done without the factory and with a few inline lambdas:

```javascript
nextMatches.filter((match, index) => filterByRouteProps(
    currentMatches, 
    prevUrl, 
    url,
    undefined,
    match, 
    index
));
```